ta/gatekeeper/include/gatekeeper_ipc.h

```diff
@@ -0,0 +1,169 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/* Copyright (c) 2018, Linaro Limited */
+/* Copyright (c) 2017, GlobalLogic  */
+
+#ifndef __GATEKEEPER_IPC_H
+#define __GATEKEEPER_IPC_H
+
+#include <stdint.h>
+#include <string.h>
+
+/*
+ * Please keep this define consistent with TA_UUID variable that defined
+ * in Android.mk file
+ */
+#define TA_GATEKEEPER_UUID { 0x4d573443, 0x6a56, 0x4272, \
+		{ 0xac, 0x6f, 0x24, 0x25, 0xaf, 0x9e, 0xf9, 0xbb} }
+
+/*
+ * GateKeeper message size
+ */
+#define RECV_BUF_SIZE		8192
+
+/*
+ * GateKeeper command identifier
+ */
+enum gatekeeper_command {
+	GK_ENROLL,
+	GK_VERIFY,
+};
+
+/*
+ * GateKeeper messages error codes
+ */
+enum gatekeeper_error {
+	GK_ERROR_NONE = 0,
+	GK_ERROR_INVALID,
+	GK_ERROR_RETRY,
+	GK_ERROR_UNKNOWN,
+};
+
+/*
+ * General message functions
+ */
+
+/*
+ * Size function. Function returns number of bytes that was used @buffer
+ * through @iter
+ *
+ * @buffer points to the start of the data
+ * @iter points to the end of read data
+ * @return number of bytes
+ */
+static inline uint32_t get_size(const uint8_t *buffer, const uint8_t *iter)
+{
+	return iter - buffer;
+}
+
+/*
+ * Serialization functions
+ */
+
+/*
+ * Blob serialization function. Function writes to @buffer first
+ * blob @length (4 bytes) and than writes blob @data (@length bytes).
+ * After function @buffer will point to next memory after written data.
+ *
+ * @buffer that will contain serialized data
+ * @data blob pointer
+ * @length blob size
+ */
+static inline void serialize_blob(uint8_t **buffer,
+				  const uint8_t *data,
+				  uint32_t length)
+{
+	memcpy(*buffer, &length, sizeof(length));
```

Comment:
From what I can see we only provide length as "sizeof(something)", so we should be safe, but if this ever gets accidentally exposed to user provided parameter, then this memcpy to potentially be dangerous without more parameter checking.

Reply:
Didn't get what you meant :(

Here we serialize to the buffer `length` variable value, and in the next line we shift `sizeof(length);` bytes

Reply:
I meant that if the `uint32_t length` would ever contain user provided size, then we have a risk of buffer overflow. But looking at the code as it is written here, it's not exposed, we take the sizeof(something) when calling this function.

Reply:
Right, if it is possible to check the length in one or another way, then I think it's worth doing that.